src/HashingLogic.ts

```diff
@@ -53,35 +120,210 @@ export const orderedStringify = (obj: {}) => {
+ * after padding, and sorting.
+ *
  */
-export const getMerkleTree = (attestations: IAttestationData[]) => {
-  const leaves = attestations.map(hashAttestation)
+export const getBloomMerkleTree = (
+  dataHashes: string[],
+  paddingNodes: string[],
+  checksumHash: string
+): MerkleTree => {
+  let leaves = dataHashes
+  leaves.push(checksumHash)
+  leaves = leaves.concat(paddingNodes)
   return getMerkleTreeFromLeaves(leaves)
 }
 
+/**
+ * Given an array of root hashes, sort and hash them into a checksum buffer
+ * @param {string[]} dataHashes - array of dataHashes as hex strings
+ */
+export const getChecksum = (dataHashes: string[]): Buffer => {
+  return ethUtil.toBuffer(hashMessage(JSON.stringify(dataHashes.sort())))
+}
+
+/**
+ * Given an array of root hashes, get and sign the checksum
+ * @param dataHashes - array of dataHashes as hex strings
+ * @param privKey - private key of signer
+ */
+export const signChecksum = (dataHashes: string[], privKey: Buffer): string => {
+  return signHash(getChecksum(dataHashes), privKey)
+}
+
+/**
+ * Sign a complete attestation node and return an object containing the datanode and the signature
+ * @param dataNode - Complete attestation data node
+ * @param globalRevocationLink - Hex string referencing revocation of the whole attestation
+ * @param privKey - Private key of signer
+ */
+export const getSignedDataNode = (
+  dataNode: IAttestation,
+  globalRevocationLink: string,
+  privKey: Buffer
+): IDataNode => {
+  const attestationNode: IAttestationNode = {
+    data: dataNode.data,
+    type: dataNode.type,
+    aux: dataNode.aux,
+    link: {
+      local: generateNonce(),
+      global: globalRevocationLink,
+    },
+  }
+  const attestationHash = hashAttestationNode(attestationNode)
+  const attestationSig = signHash(attestationHash, privKey)
+  return {
+    attestationNode: attestationNode,
+    signedAttestation: attestationSig,
+  }
+}
+
+/**
+ * Given the number of data nodes return an array of padding nodes
+ * @param {number} dataCount - number of data nodes in tree
+ *
+ * A Bloom Merkle tree will contain at minimum one data node and one checksum node
+ * In order to obscure the amount of data in the tree, the number of nodes are padded to
+ * a set threshold
+ *
+ * The Depth of the tree increments in steps of 5
+ * The number of terminal nodes in a filled binary tree is 2 ^ (n - 1) where n is the depth
+ *
+ * dataCount 1 -> 15: paddingCount: 14 -> 0 (remeber + 1 for checksum node)
+ * dataCount 16 -> 511: paddingCount 495 -> 0
+ * dataCount 512 -> ...: paddingCount 15871 -> ...
+ * ...
+ */
+export const getPadding = (dataCount: number): string[] => {
+  if (dataCount < 1) return []
+  let i = 5
+  while (dataCount + 1 > 2 ** (i - 1)) {
+    i += 5
+  }
+  const paddingCount = 2 ** (i - 1) - (dataCount + 1)
+  return Array.apply(null, Array(paddingCount)).map(
+    (item: number, index: number) => {
+      return hashMessage(crypto.randomBytes(20))
```

Comment:
output of crypto.randombytes should be stored so the attester knows he's not being spoofed

Reply:
spoofed meaning that a padding node could be a disguised data node? I think that's handled by using the checksum node and the fact that each node is signed